### ISSUE TITLE
CA1512 is raised for types that do not satisfy generic constraint

### ISSUE THREAD

**Issue by h3xds1nz** [dotnet/roslyn-analyzers#7617]
### Analyzer

**Diagnostic ID**: [CA1512](https://learn.microsoft.com/en-us/dotnet/fundamentals/code-analysis/quality-rules/ca1512): `Use ArgumentOutOfRangeException throw helper`

### Describe the bug

`CA1512` currently does not care whether the type in question implements the constraint of the method (e.g. `IEquatable<T>, IComparable<T>, INumberBase<T>`), so it will offer code-fixes that cannot possibly work.

### Steps To Reproduce

```csharp
private void BadExample(BadArgument argument, BadArgument argument2)
{
    // Use any overriden operator to trigger CA1512
    if (argument == argument2)
        throw new ArgumentOutOfRangeException();
}

public readonly struct BadArgument
{
  public static bool operator ==(BadArgument a, BadArgument b) => true;
  public static bool operator !=(BadArgument a, BadArgument b) => true;
  public static bool operator >(BadArgument a, int b) => true;
  public static bool operator <(BadArgument a, int b) => true;
}
```

### Expected behavior

Analyzer is not flagging CA1512 for types that do not implement the interface required for the given method.

### Actual behavior

While you might expect e.g. `IEquatable<T>` or `IComparable<T>` implemented when operators are overriden, this is often not true on e.g. public .NET APIs, so it creates unnecessary noise that must be suppressed by the user.
